source/common/http/conn_manager_impl.cc

```diff
@@ -1631,7 +1631,9 @@ void ConnectionManagerImpl::ActiveStream::encodeHeaders(ActiveStreamEncoderFilte
 void ConnectionManagerImpl::ActiveStream::encodeHeadersInternal(ResponseHeaderMap& headers,
                                                                 bool end_stream) {
   // Base headers.
-  connection_manager_.config_.dateProvider().setDateHeader(headers);
+  if (!connection_manager_.config_.shouldPreserveUpstreamDate() || !headers.Date()) {
```

Comment:
This PR looks great.
My only caveat is that I believe this will result in also preserving the date header when we serve out of cache.
Is your intent to do this (since at the end of the day the source of that date is upstream) or is the intent to only overwrite date headers served live from upstream (preserve date headers when they're relevant, but let cache results have an accurate up-to-date date)?  I'd lean towards the latter, since I think that's a more clear interpretation of the config field, at which point I think you have to differentiate between the two.

Reply:
I think the intention here would be to use a new date for serving cached content. Otherwise, we may end up with stale response dates that might be confusing to downstream components (e.g. the client).

Is there a convention around how to determine whether a cached response is being served?

Reply:
Yeah, I was looking into that, and there's no an obvious direct check.

You can infer from the rc details being "via_upstream" but that's a string compare and icky.  You can infer from the stream info upstreamHost being set, which is probably the easiest workaround. 
I'd also be fine with something added to stream info to indicate the source of the reply - cache / Envoy generated / proxied from upstream.  @mattklein123 any thoughts here?

Reply:
cc @jmarantz @toddmgreer

Reply:
Yeah I would think for the caching case we might want to introduce a stream info flag. This seems useful/required anyway for logging purposes in a production ready caching solution.

Reply:
SGTM.
@chradcliffe do you want to do that in this PR, or do you want me to throw something together to that end (probably tomorrow)?

Reply:
I can give it a try in this PR -- the "stream info flag" in this case would be a new value in `StreamInfo::ResponseFlag`, right?

Reply:
You got it!

Reply:
Responses served from cache should have the same date header as when they got inserted into cache. The date header is defined to represent the time the message was originated, regardless of whether the message happens to spend some time in a cache.

Reply:
Oh interesting.
That's not what's happening today.  So in that case this PR LGTM, but someone should do the inverse and not set the date header if it passed through the cache filter.  I'm happy to pick that up since I noticed the caching code fails to set rc details

Reply:
> The date header is defined to represent the time the message was originated, regardless of whether the message happens to spend some time in a cache.

Interesting. I guess Envoy's default behavior is wrong then? In any case, I think we can address this out of the scope of this PR per other discussion.